### ISSUE TITLE
How to use SSD to train and test  images with only one channel?

### ISSUE THREAD

**Issue by mjssat7** [weiliu89/caffe#529]
My image datasets are all gray images,they all only have one channel,not three channels. I use SSD to do training and testing on my datasets. I also use caffe to do it, but I have one problem ,for example:

I0411 22:39:47.866865 14814 layer_factory.hpp:77] Creating layer data
I0411 22:39:47.867352 14814 net.cpp:100] Creating Layer data
I0411 22:39:47.867369 14814 net.cpp:408] data -> data
I0411 22:39:47.867410 14814 net.cpp:408] data -> label
I0411 22:39:47.868923 14824 db_lmdb.cpp:35] Opened lmdb examples/MINISAR/MSRR/MSRR_trainval_lmdb
I0411 22:39:47.887765 14814 annotated_data_layer.cpp:62] output data size: 4,1,300,300
I0411 22:39:47.894155 14814 net.cpp:150] Setting up data
I0411 22:39:47.894176 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894183 14814 net.cpp:157] Top shape: 1 1 1 8 (8)
I0411 22:39:47.894187 14814 net.cpp:165] Memory required for data: 1440032
I0411 22:39:47.894197 14814 layer_factory.hpp:77] Creating layer data_data_0_split
I0411 22:39:47.894213 14814 net.cpp:100] Creating Layer data_data_0_split
I0411 22:39:47.894222 14814 net.cpp:434] data_data_0_split <- data
I0411 22:39:47.894237 14814 net.cpp:408] data_data_0_split -> data_data_0_split_0
I0411 22:39:47.894251 14814 net.cpp:408] data_data_0_split -> data_data_0_split_1
I0411 22:39:47.894261 14814 net.cpp:408] data_data_0_split -> data_data_0_split_2
I0411 22:39:47.894271 14814 net.cpp:408] data_data_0_split -> data_data_0_split_3
I0411 22:39:47.894281 14814 net.cpp:408] data_data_0_split -> data_data_0_split_4
I0411 22:39:47.894289 14814 net.cpp:408] data_data_0_split -> data_data_0_split_5
I0411 22:39:47.894297 14814 net.cpp:408] data_data_0_split -> data_data_0_split_6
I0411 22:39:47.894433 14814 net.cpp:150] Setting up data_data_0_split
I0411 22:39:47.894443 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894450 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894456 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894462 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894469 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894474 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894481 14814 net.cpp:157] Top shape: 4 1 300 300 (360000)
I0411 22:39:47.894486 14814 net.cpp:165] Memory required for data: 11520032
I0411 22:39:47.894491 14814 layer_factory.hpp:77] Creating layer conv1
I0411 22:39:47.894510 14814 net.cpp:100] Creating Layer conv1
I0411 22:39:47.894517 14814 net.cpp:434] conv1 <- data_data_0_split_0
I0411 22:39:47.894526 14814 net.cpp:408] conv1 -> conv1
OpenCV Error: Assertion failed ((scn == 3 || scn == 4) && (depth == CV_8U || depth == CV_32F)) in cvtColor, file /home/dell1714/NVIDIA/opencv/opencv-3.0.0/modules/imgproc/src/color.cpp, line 7946
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/dell1714/NVIDIA/opencv/opencv-3.0.0/modules/imgproc/src/color.cpp:7946: error: (-215) (scn == 3 || scn == 4) && (depth == CV_8U || depth == CV_32F) in function cvtColor

*** Aborted at 1491921589 (unix time) try "date -d @1491921589" if you are using GNU date ***
PC: @     0x7f6453f38418 (unknown)
*** SIGABRT (@0x3e8000039de) received by PID 14814 (TID 0x7f6424b43700) from PID 14814; stack trace: ***
    @     0x7f6453f384a0 (unknown)
    @     0x7f6453f38418 (unknown)
    @     0x7f6453f3a01a (unknown)
    @     0x7f6454d4b84d (unknown)
    @     0x7f6454d496b6 (unknown)
    @     0x7f6454d49701 (unknown)
    @     0x7f6454d49919 (unknown)
    @     0x7f644b844582 cv::error()
    @     0x7f644b844713 cv::error()
    @     0x7f644a667a68 cv::cvtColor()
    @     0x7f6456059018 caffe::AdjustSaturation()
    @     0x7f645605e14b caffe::RandomSaturation()
    @     0x7f645605e90f caffe::ApplyDistort()
    @     0x7f645606db92 caffe::DataTransformer<>::DistortImage()
    @     0x7f6455eedb86 caffe::AnnotatedDataLayer<>::load_batch()
    @     0x7f6455f8e20f caffe::BasePrefetchingDataLayer<>::InternalThreadEntry()
    @     0x7f6455fc9385 caffe::InternalThread::entry()
    @     0x7f6449d915d5 (unknown)
    @     0x7f64445b66fa start_thread
    @     0x7f6454009b5d (unknown)
Aborted (core dumped)
So how to solve this problem, and I know that setting force_color: true  is a solution,but it is not good for me, Then what else I can do to deal with this problem, can you help me?@weiliu ,thanks!

**Comment by weiliu89** [comment]
Set force_color to true in the transform_param.

**Comment by mjssat7** [comment]
Thank you very much @weiliu, When I set force_color to true, then it will like this to be 3 channels, but I only want it to be 1 channel, so what else can I  do to solve this problem. Does SSD only support traing on images with 3 channels ???  Thanks  a lot ! @ weiliu
I0411 22:39:47.887765 14814 annotated_data_layer.cpp:62] output data size: 4,3,300,300
I0411 22:39:47.894155 14814 net.cpp:150] Setting up data
I0411 22:39:47.894176 14814 net.cpp:157] Top shape: 4 3 300 300 (360000)
I0411 22:39:47.894183 14814 net.cpp:157] Top shape: 1 1 1 8 (8)

**Comment by weiliu89** [comment]
It would only replicate the gray channel three times to get a three channel image. It eventually look the same as a single channel gray image. You can check how it looks like in OpenCV by Gray2RGB.

The reason is because we are taking a ImageNet pretrained model, which requires the input with three channels.

**Comment by mjssat7** [comment]
Thanks a lot !@weiliu89, because I have a pretrained model trained on my own image datasets,which all have one channel, so it requires the input with one channel. Thus I have to finds solutions to deal with this problem above, and I have tried so many times to modify SSD code , but I'm failed, so I hope that you could  give me some suggestions to solve this problem, thank you very much ! I'm sorry to disturb you at this time. @weiliu89

**Comment by weiliu89** [comment]
I think you should turn off all the color distortion augmentation.

**Comment by mjssat7** [comment]
Yes, you are right, I just turn off  all the color distortion augmentation ,and it works ! Thank you very much!
One more problem, as use gray images, should I  set mean_value or not? I think maybe not. thanks @weiliu89

**Comment by JPBoost** [comment]
hi @mjssat7 
   I have the same problem with you. How do you turn off the color distortion to make it work? Can you tell me some more ?
   thank you very much!
